internal/metric/registry/registry.go

```diff
@@ -122,18 +100,28 @@ func (u *UniqueInstrumentMeterImpl) NewAsyncInstrument(
 	u.lock.Lock()
 	defer u.lock.Unlock()
 
-	impl, err := u.checkUniqueness(descriptor)
-
-	if err != nil {
-		return nil, err
-	} else if impl != nil {
-		return impl.(sdkapi.AsyncImpl), nil
+	impl, ok := u.state[descriptor.Name()]
+	if !ok {
+		asyncInst, err := u.impl.NewAsyncInstrument(descriptor, runner)
+		if err != nil {
+			return nil, err
+		}
+		u.state[descriptor.Name()] = asyncInst
+		return asyncInst, nil
 	}
 
-	asyncInst, err := u.impl.NewAsyncInstrument(descriptor, runner)
-	if err != nil {
-		return nil, err
+	// Return an ErrMetricKindMismatch error if there is a conflict between
+	// a descriptor that was already registered and the `descriptor` argument
+	if !compatible(descriptor, impl.Descriptor()) {
+		return nil, NewMetricKindMismatchError(impl.Descriptor())
 	}
-	u.state[descriptor.Name()] = asyncInst
-	return asyncInst, nil
+
+	return impl.(sdkapi.AsyncImpl), nil
```

Comment:
This will panic if an Async instrument is registered with the same name.

While the original implementation suffered from this also I think now would be a chance to fix.

Reply:
Not a bug, see "compatible" which checks for instrument kind, also see #2460